server/pulp/plugins/rsync/publish.py

```diff
@@ -0,0 +1,443 @@
+
+        :param files_from: absolute path to a file with list of file paths to rsync
+        :type files_from: str
+        :param source_prefix: path to directory that contains all paths in files_from
+        :type source_prefix: str
+        :param dest_prefix: path to directory on remote server where files should be rsynced
+        :type dest_prefix: str
+        :param exclude: list of file/directory paths to exclude
+        :type exclude: list
+        :return: list of arguments for rsync command
+        :rtype: list of strings
+        """
+        args = [self._CMD, self._AZQ]
+        if not self.delete:
+            args.extend(["--files-from", files_from, "--relative"])
+
+        if exclude:
+            for x in exclude:
+                args.extend(["--exclude", x])
+        args.extend(self.make_authentication())
+        if self.delete:
+            args.append("--delete")
+        if self.links:
+            args.append("--links")
+        else:
+            args.append("--copy-links")
+        args.append(source_prefix)
+        args.append(self.make_destination(dest_prefix))
+        return args
+
+    def rsync(self):
+        """
+        This method formulates the rsync command based on parameters passed in to the __init__ and
+        then executes it.
+
+        :return: (boolean indicating success or failure, stdout + stderr)
+        :rtype: tuple
+        """
+        if not self.file_list:
+            return (True, "Nothing to sync")
+        output = ""
+        list_of_files = os.path.join(self.get_working_dir(), str(uuid.uuid4()))
+        open(list_of_files, 'w').write("\n".join(sorted(self.file_list)))
+
+        # copy files here, not symlinks
+        (is_successful, this_output) = self.remote_mkdir(self.dest_directory)
+        if not is_successful:
+            _logger.error("Cannot create directory %s: %s" % (self.dest_directory, this_output))
+            return (is_successful, this_output)
+        output += this_output
+        rsync_args = self.make_rsync_args(list_of_files, self.src_directory,
+                                          self.dest_directory, self.exclude)
+        (is_successful, this_output) = self.call(rsync_args)
+        _logger.info(this_output)
+        if not is_successful:
+            _logger.error(this_output)
+            return (is_successful, this_output)
+        output += this_output
+        return (is_successful, output)
+
+    def process_main(self):
+        (successful, output) = self.rsync()
+        if not successful:
+            raise PulpCodedException(message=output)
+
+
+class Publisher(PublishStep):
+    """
+    RSync publisher class that provides the common code for publishing to remote server. Each
+    plugin needs to extend this class and add necesary logic to the _add_necesary_steps method.
+
+    :ivar symlink_list: list of symlinks to rsync
+    :ivar content_unit_file_list: list of content units to rsync
```

Comment:
There are a lot more variables than this. Some non-public ones could have `_` put in front of them. Also this could be consolidated with the variables in the docstring for `__init__()`.